packages/vite/src/node/build.ts

```diff
@@ -1548,12 +1549,6 @@ export interface BuilderOptions {
   buildApp?: (builder: ViteBuilder) => Promise<void>
 }
 
-async function defaultBuildApp(builder: ViteBuilder): Promise<void> {
-  for (const environment of Object.values(builder.environments)) {
-    await builder.build(environment)
-  }
-}
```

Comment:
Instead of https://github.com/vitejs/vite/commit/dcea0b1a20521088a9afcb1322a725ce2c04596e, can the default behavior adjusted by checking `environment.isBuilt` for each environment here?

```js
if (!environment.isBuilt) {
  await builder.build(environment)
}
```

I think the difference is whether `plugin.buildApp` partially building environments should cause Vite to build the rest of environments. As a default behavior, this seemed natural to me. `plugin.buildApp` can still technically avoid that by doing `delete builder.environments["noNeedThisEnv"]` :thinking:

Reply:
We can't do that except if we move the order post hooks before config.builder.buildApp, and we also may deprecate and remove config.builder.buildApp in the future